runtime/src/crowdfund.rs

```diff
@@ -0,0 +1,539 @@
+		/// slot. It will be withdrawable in two instances: the parachain becomes retired; or the
+		/// slot is
+		fn contribute(origin, #[compact] index: FundIndex, #[compact] value: BalanceOf<T>)
+		{
+			let who = ensure_signed(origin)?;
+
+			ensure!(value >= T::MinContribution::get(), "contribution too small");
+
+			let mut fund = Self::funds(index).ok_or("invalid fund index")?;
+			fund.raised += value;
+			ensure!(fund.raised <= fund.cap, "contributions exceed cap");
+			let now = <system::Module<T>>::block_number();
+			ensure!(fund.end > now, "contribution period ended");
+
+			T::Currency::transfer(&who, &Self::fund_account_id(index), value)?;
+
+			let id = Self::id_from_index(index);
+			let balance = who.using_encoded(|b| child::get_or_default::<BalanceOf<T>>(id.as_ref(), b));
+			let balance = balance.saturating_add(value);
+			who.using_encoded(|b| child::put(id.as_ref(), b, &balance));
+
+			if <slots::Module<T>>::is_ending(now).is_some() {
+				// Now in end period; record it
+				fund.last_contribution = Some(now);
+				if let Some(c) = fund.last_contribution {
+					if c != now {
+						// last contribution was at earlier time; re-insert into `NewRaise`
+						NewRaise::mutate(|v| v.push(index));
+					}
+				}
+			} else {
+				if fund.last_contribution.is_some() {
+					// Now outside end period and last was also inside end period: reset last to
+					// None and insert (because it will have been removed at end of last block).
+					fund.last_contribution = None;
+					NewRaise::mutate(|v| v.push(index));
+				}
+			}
+
+			<Funds<T>>::insert(index, &fund);
+
+			Self::deposit_event(RawEvent::Contributed(who, index, value));
+		}
+
+		
+		/// Withdraw full balance of a contributor to an unsuccessful fund.
+		fn withdraw(origin, #[compact] index: FundIndex) {
+			let who = ensure_signed(origin)?;
+
+			let mut fund = Self::funds(index).ok_or("invalid fund index")?;
+			let now = <system::Module<T>>::block_number();
+			ensure!(now >= fund.end, "contribution period not over");
+
+			let id = Self::id_from_index(index);
+			let balance = who.using_encoded(|b| child::get::<BalanceOf<T>>(id.as_ref(), b))
+				.ok_or("not a contributor")?;
+
+			// Avoid using transfer to ensure we don't pay any fees.
+			let _ = T::Currency::resolve_into_existing(&who, T::Currency::withdraw(
+				&Self::fund_account_id(index),
+				balance,
+				WithdrawReason::Transfer,
+				ExistenceRequirement::AllowDeath
+			)?);
+
+			let id = Self::id_from_index(index);
+			who.using_encoded(|b| child::kill(id.as_ref(), b));
+			fund.raised = fund.raised.saturating_sub(balance);
+
+			<Funds<T>>::insert(index, &fund);
+
+			Self::deposit_event(RawEvent::Withdrew(who, index, balance));
+		}
+		
+		/// Note that a successful fund has lost its parachain slot, and place it into retirement.
+		fn begin_retirement(_origin, #[compact] index: FundIndex) {
+			// origin unimportant.
```

Comment:
When you say "origin unimportant", I think you mean "allow for any origin".

But I think we should explicitly check that it is **not** inherent origin, else a malicious actor could spam the chain with function calls, causing storage lookup and other logic.

What do you think?

Reply:
sure we can `ensure_signed`.

Reply:
I have updated all functions which did not use origin to have `ensure_signed`.